File: week_1/day-5-string-hashing/valid_anagram_follow_up_questions/isAnagram.py
def isAnagram(s, t):
    if len(s) != len(t):
        return False

    freq = [0] * 26  # only for 'a' to 'z'

    for ch in s:
        freq[ord(ch) - ord('a')] += 1

    for ch in t:
        freq[ord(ch) - ord('a')] -= 1
        if freq[ord(ch) - ord('a')] < 0:
            return False

    return True


s = "listen"
t = "silent"

print("Are the strings anagrams?", isAnagram(s, t))

File: week_1/day-5-string-hashing/valid_anagram_follow_up_questions/isAnagram.js
function isAnagram(s, t) {
    if (s.length !== t.length) return false;

    const freq = new Array(26).fill(0);

    for (let ch of s) {
        freq[ch.charCodeAt(0) - 97]++;
    }

    for (let ch of t) {
        freq[ch.charCodeAt(0) - 97]--;
        if (freq[ch.charCodeAt(0) - 97] < 0) return false;
    }

    return true;
}

const s = "listen";
const t = "silent";

console.log("Are the strings anagrams?", isAnagram(s, t));
